### ISSUE TITLE
Arcus.Security.Core.SecretNotFoundException: 'The secret SyncFusionLicense was not found.'

### ISSUE THREAD

**Issue by MarkLFT** [arcus-azure/arcus.security#169]
I am trying to use your KeyVaultSecret provider to access the Azure key vault.  Like many others I keep my licenses for purchased 3rd party libraries there.

I have tried using both the CertificateBasedAuthentication and the ServicePrincipalAuthentication, but in both cases when trying to read the secret value I receive an Arcus.Security.Core.SecretNotFoundException: 'The secret SyncFusionLicense was not found.' error.

The name of the secret is very simple, no spaces, dashes etc. and it is spelled correctly.  I believe the authentication is setup correctly on the AKV as I previously used the AddAzureKeyVault method, and it worked, well that was until they introduced a file version error in the ActiveDirectory dependencies.

My question is, is there are special formatting I would need for the secrets name?   And would an authentication error for example give  this error, or is it guaranteed to be an incorrect spelling of the secrets name?

Here is the code I am using to configured the secrets provider.

```csharp
.ConfigureSecretStore((hostingContext, config, builder) =>
                    {
                        var keyVaultUri = $"https://{config["KeyVault:KeyVaultName"]}.vault.azure.net/";
                        var vaultAuthenticator = new ServicePrincipalAuthentication(config["KeyVault:AzureADApplicationId"], config["KeyVault:AzureADSecret"]);
                        var vaultConfiguration = new KeyVaultConfiguration(keyVaultUri);
                        var keyVaultSecretProvider = new KeyVaultSecretProvider(vaultAuthenticator, vaultConfiguration);
                        builder.AddProvider(keyVaultSecretProvider);
```

and here is the code I am using to retrieve the secret value

```csharp
public async void Configure(IApplicationBuilder app, IWebHostEnvironment env, ISecretProvider secretsProvider)
        {
            var license = await secretsProvider.GetSecretAsync("SyncFusionLicense");
            Syncfusion.Licensing.SyncfusionLicenseProvider.RegisterLicense(license.Value);
```

Can you please explain what I am doing wrong.

Many thanks.

**Comment by tomkerkhove** [comment]
First of all a tip for you! You can simply setting up the secret provider to just a few lines ([docs](https://security.arcus-azure.net/features/secret-store/provider/key-vault)).

I think everything is setup but the main issue here is that the secret you are looking for is not found. So I'd suggest checking your Azure Key Vault instance and make sure that it's there.

Can you do that please?

**Comment by MarkLFT** [comment]
I have checked it is there,  I copied and pasted the name to be 100% certain.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by tomkerkhove** [comment]
In that case, can you share some logs, please? Otherwise, we won't be able to help further because everything looks fine and this exception is only thrown when the secret providers don't find a secret.

**Comment by MarkLFT** [comment]
Apologies for my delay in responding, I have been undertaking  some additional testing using different scenarios to try to find the cause of this.

Firstly I focused on using the a secret instead of a certificate to authenticate, and after checking my code I found a typo in the authentication password.  This gave me access to the secrets, yes the secret name was correct and the secret was in the vault.  So the issue with this is you library seems to report secret not found, no matter the cause of the problem.  This is not helpful.  If it had reported an authentication error it would have been far more useful.

Secondly I have been testing using the certificate for authentication.  In case there was mismatch in the certificate or some other similar issue, I created a new certificate, and uploaded it to Azure, and onto my computer, still no joy.  The problem is, as above, it is only reporting as being unable to locate the secret, and not the real cause.  I have searched through all the logs I can find, and everywhere only reports secret not found.   Is there somewhere I can look that might give me a clue about the true cause or error message?

The reason I tried your libraries to manage secrets was there is an issue with the MS solution using AddAzureKeyVault 
 with certificates, this is well documented in their issues.  I am wondering if you are using their libraries underneath, and that is why I cannot get the certificates to authenticate also.

I do not particularly want to store secrets either in the code, or in configuration files that we install on clients computers, hence the use of certificates, so I really need to find a way to get certificate authentication to work.  Any advise you can give that can point me in the right direction would be helpful.

**Comment by tomkerkhove** [comment]
> Firstly I focused on using the a secret instead of a certificate to authenticate, and after checking my code I found a typo in the authentication password. This gave me access to the secrets, yes the secret name was correct and the secret was in the vault. So the issue with this is you library seems to report secret not found, no matter the cause of the problem. This is not helpful. If it had reported an authentication error it would have been far more useful.

For the record, it's not my library but it's open to everyone 😄 But you have a point here, this should not be the case and instead throw a dedicated exception related to auth. @stijnmoreels can you look into this please? This is indeed very confusing.

> Secondly I have been testing using the certificate for authentication. In case there was mismatch in the certificate or some other similar issue, I created a new certificate, and uploaded it to Azure, and onto my computer, still no joy. The problem is, as above, it is only reporting as being unable to locate the secret, and not the real cause. I have searched through all the logs I can find, and everywhere only reports secret not found. Is there somewhere I can look that might give me a clue about the true cause or error message?
> 
> The reason I tried your libraries to manage secrets was there is an issue with the MS solution using AddAzureKeyVault
> with certificates, this is well documented in their issues. I am wondering if you are using their libraries underneath, and that is why I cannot get the certificates to authenticate also.

As far as I remember we are just extending the Azure SDK, but we are not using `AddAzureKeyVault` as we are not building on top of IConfiguration (as [they serve different masters](https://blog.tomkerkhove.be/2020/06/22/introducing-secret-store-making-secrets-a-first-class-citizen-in-net-core/))
 
> I do not particularly want to store secrets either in the code, or in configuration files that we install on clients computers, hence the use of certificates, so I really need to find a way to get certificate authentication to work. Any advise you can give that can point me in the right direction would be helpful.

That's a good motto and instead of using certificates you can even use Azure Managed Identity so that you don't have *any* secrets! You can learn more about it [in our docs](https://security.arcus-azure.net/features/auth/azure-key-vault) but do note that this requires to run in Azure. THere are some tricks to run Managed Identity locally but I'm not an expert on that front as I use service principles for local development.

Hope this helps you further and sorry for the pain.

**Comment by MarkLFT** [comment]
> As far as I remember we are just extending the Azure SDK, but we are not using AddAzureKeyVault as we are not building on top of IConfiguration (as they serve different masters)

From what I can read the issue with certificate authentication is not related directly to the AddAzureKeyVault, but more to do with it's AD Authentication dependencies, maybe the same dependencies you are using in the Azure SDK?

> That's a good motto and instead of using certificates you can even use Azure Managed Identity so that you don't have any secrets! You can learn more about it in our docs but do note that this requires to run in Azure. There are some tricks to run Managed Identity locally but I'm not an expert on that front as I use service principles for local development.

We cannot host these services in Azure, as each one of our customers have these services running on services in their properties.    They are simple ASP .NET Core API Services.  So my only real options, are secrets or certificates, of the two, certificates are the safest.

I guess I am going to have to look for another secret store I can use, one that is not related to Azure or it's SDK's.

**Comment by tomkerkhove** [comment]
For what it's worth, we have HashiCorp vault coming up; if that helps.

**Comment by tomkerkhove** [comment]
Based on #170 it looks like it works as expected, but maybe you are running an older version. Can you share that please @MarkLFT?

**Comment by stijnmoreels** [comment]
@MarkLFT , we have added a feature to our secret store to configure the critical exceptions of the secret provider yourself. Within this change, we also updated our Azure Key Vault secret provider to include any authentication and authorization exceptions as this critical exceptions so the next time one of those is thrown, the secret store will behave differently.

Have a look at our currently [preview docs](https://security.arcus-azure.net/preview/features/secret-store/create-new-secret-provider#add-critical-exceptions) for more information and keep an eye on the next release of the library to use this feature. Thanks for the issue tracking!
